src/sql/base/browser/ui/chart/chart.component.ts

```diff
@@ -0,0 +1,321 @@
+})
+export class Chart<TChartType extends azdata.ChartType, TData extends azdata.ChartData<TChartType>, TOptions extends azdata.ChartOptions<TChartType>> extends Disposable {
+	private _type: TChartType;
+	private _data: chartjs.ChartData;
+
+	private chart: chartjs.Chart;
+	private canvas: HTMLCanvasElement;
+	private chartCanvasId: string;
+
+	/**
+	 * Options in the form that Chart.js accepts
+	 */
+	private _options: chartjs.ChartOptions = {
+		events: ['click', 'keyup'],
+		responsive: true,
+		maintainAspectRatio: false
+	};
+
+	constructor(
+		@Inject(forwardRef(() => ChangeDetectorRef)) private _changeRef: ChangeDetectorRef
+	) {
+		chartjs.Chart.register(...chartjs.registerables);
+		super();
+
+		this.chartCanvasId = 'chart' + generateUuid();
+	}
+
+	ngAfterViewInit(): void {
+
+	}
+
+	/**
+	 * Setter function for chart type
+	 */
+	public set type(val: any) {
+		if (val === 'horizontalBar') {
+			// In Chart.js, horizontal bar charts are just bar charts with a different indexAxis set.
+			// The indexAxis gets set here, and the Chart.js type gets mapped at conversion time.
+			this._options = mixin({}, mixin(this._options, { indexAxis: 'y' }));
+		}
+		else {
+			this._type = val;
+		}
+		this._changeRef.detectChanges();
+	}
+
+	/**
+	 * Setter function for chart data
+	 */
+	public set data(val: TData) {
+		this._data = this.convert(val);
+
+		this.drawChart();
+	}
+
+	/**
+	 * Setter function for chart options.
+	 * Some options like responsiveness and maintainaspectratio are set by default and will be used even if no options are provided.
+	 */
+	public set options(val: TOptions) {
+		if (val === undefined) {
+			return;
+		}
+
+		// Free-form chart.js options get added first...
+		this._options = mixin({}, mixin(this._options, val.freeformOptions));
+
+		// ...then strongly-typed ComponentModel options get set (overriding free-form options)
+		if (val !== undefined) {
+			if (val.chartTitle) { // undefined results in hiding title
+				if (typeof val.chartTitle === 'string') {
+					this._options = mixin(this._options, {
+						plugins: {
+							title: {
+								text: val.chartTitle,
+								display: true
+							}
+						}
+					});
+				}
+			} else {
+				this._options = mixin(this._options, { plugins: { title: { display: false } } });
+			}
+
+			if (val.legendVisible !== false) { // undefined defaults to true
+				this._options = mixin(this._options, { plugins: { legend: { display: true } } });
+			} else {
+				this._options = mixin(this._options, { plugins: { legend: { display: false } } });
+			}
+		}
+
+		this.drawChart();
+	}
+
+	public set height(val: string | number) {
+		if (val && this.chart) {
+			(this.chart.canvas.parentNode as any).style.height = val;
+		}
+	}
+
+	public set width(val: string | number) {
+		if (val && this.chart) {
+			(this.chart.canvas.parentNode as any).style.width = val;
+		}
+	}
+
+	private convert(val: azdata.ChartData<TChartType>): chartjs.ChartData {
```

Comment:
nit: name could be clearer - something like `convertChartDataToChartJsData` or at least `convertData` (with doc comment)

Reply:
Yup, good call; now that this is no longer the only conversion function, it requires a better name.